compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java

```diff
@@ -905,17 +907,61 @@ public void visit(BLangArrayType arrayTypeNode) {
-                        (size == OPEN_SEALED_ARRAY_INDICATOR) ?
-                                new BArrayType(resultType, arrayTypeSymbol, size, BArrayState.OPEN_SEALED) :
-                                new BArrayType(resultType, arrayTypeSymbol, size, BArrayState.CLOSED_SEALED);
-            }
-            resultType = arrayTypeSymbol.type = arrType;
+                BLangExpression size = arrayTypeNode.sizes[i];
+                if (size instanceof BLangLiteral) {
+                    arrType = ((Integer) (((BLangLiteral) size).getValue()) == UNSEALED_ARRAY_INDICATOR) ?
+                            new BArrayType(resultType, arrayTypeSymbol,
+                                    (Integer) (((BLangLiteral) size).getValue()),
+                                    BArrayState.UNSEALED) :
+                            ((Integer) (((BLangLiteral) size).getValue()) == OPEN_SEALED_ARRAY_INDICATOR) ?
+                                    new BArrayType(resultType, arrayTypeSymbol,
+                                            (Integer) (((BLangLiteral) size).getValue()),
+                                            BArrayState.OPEN_SEALED) :
+                                    new BArrayType(resultType, arrayTypeSymbol,
+                                            (Integer) (((BLangLiteral) size).getValue()),
+                                            BArrayState.CLOSED_SEALED);
+                    arrayTypeSymbol.type = arrType;
+                    markParameterizedType(arrType, arrType.eType);
+                    resultType = arrayTypeSymbol.type;
 
-            markParameterizedType(arrType, arrType.eType);
+                } else {
+                    Name pkgAlias = names.fromIdNode(((BLangSimpleVarRef) size).pkgAlias);
+                    Name typeName = names.fromIdNode(((BLangSimpleVarRef) size).variableName);
+
+                    BSymbol sizeSymbol = lookupMainSpaceSymbolInPackage(size.pos, env, pkgAlias, typeName);
+
+                    if (!(symTable.notFoundSymbol.equals(sizeSymbol))) {
+                        if (sizeSymbol.tag == SymTag.CONSTANT) {
+                            BType lengthLiteralType = ((BConstantSymbol) sizeSymbol).literalType;
+                            if (lengthLiteralType.tag == TypeTags.INT) {
+                                long length = (long) ((BConstantSymbol) sizeSymbol).value.value;
+                                arrType = new BArrayType(resultType, arrayTypeSymbol,
+                                        (int) length,
+                                        BArrayState.CLOSED_SEALED);
+                                arrayTypeSymbol.type = arrType;
+                                markParameterizedType(arrType, arrType.eType);
+                                resultType = arrayTypeSymbol.type;
+
+                            } else {
+                                dlog.error(arrayTypeNode.pos,
+                                        DiagnosticCode.INVALID_ARRAY_SIZE_REFERENCE_TYPE,
```

Comment:
I think we can reuse existing error messages for these errors. For example here we can say

`incompatible types: expected 'int', found 'T'` where T is the actual type.